### PR TITLE
Fix compiling with PETSc with complex scalar type

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes https://github.com/dealii/dealii/issues/15847.

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by tamiko** [comment]
I have to get the complex number type test variant back online :-(

**Review by tamiko — APPROVED** [review]

File: source/lac/petsc_vector_base.cc [hunk 1]
```diff
@@ -277,8 +277,8 @@ namespace PETScWrappers
         // numbers back as integers later on, we get the same thing.
         for (PetscInt i = 0; i < end_index - ghost_start_index; i++)
           {
-            Assert(static_cast<PetscInt>(static_cast<PetscScalar>(
-                     ghost_start_index + i)) == (ghost_start_index + i),
+            Assert(static_cast<PetscInt>(std::real(static_cast<PetscScalar>(
```
Comment (stefanozampini) [inline source/lac/petsc_vector_base.cc:280]: Isn't it missing an array[...] somewhere?

Comment (stefanozampini) [reply source/lac/petsc_vector_base.cc:280]: ah, now I see why this was done. Ok for me

**Comment by bangerth** [comment]
Does this need to go on the release branch too? Or did we add the assertion after the release?

**Comment by masterleinad** [comment]
> Does this need to go on the release branch too? Or did we add the assertion after the release?

It was introduced in https://github.com/dealii/dealii/commit/4e27e9bf747 which is part of the release.

**Pull request merged**